### PR TITLE
Added bindings to JavascriptEditorViewport

### PR DESCRIPTION
**Pull request by arthursw**
Added bindings to `JavascriptEditorViewport` to be able to:
- react to mouse moves and inputs, and keyboard inputs,
- set the viewport type and the view mode,
- project/unproject from screen to world.

Also added a hack to be able to simply draw a selection rectangle on the viewport canvas in `FJavascriptEditorViewportClient::Draw(FViewport* Viewport, FCanvas* Canvas)`, with `SelectionBeginPoint` and `SelectionEndPoint`. This should be improved, but it would probably mean exposing FCanvas and FCanvasBoxItem to Blueprint/Javascript.

### CODE REVIEW

File: Source/JavascriptEditor/JavascriptEditorViewport.cpp [hunk 1]
```diff
@@ -70,6 +124,19 @@ class FJavascriptEditorViewportClient : public FEditorViewportClient
...
+	virtual void Draw(FViewport* Viewport, FCanvas* Canvas) override
+	{
+		FEditorViewportClient::Draw(Viewport, Canvas);
+
+		if (Widget->SelectionBeginPoint.X >= 0)
+		{
+			FCanvasBoxItem Box(Widget->SelectionBeginPoint, Widget->SelectionEndPoint - Widget->SelectionBeginPoint);
+			Box.SetColor(FColor::White);
+			Box.LineThickness = 1.0;
+			Box.Draw(Canvas);
```
Comment (nakosung) [inline Source/JavascriptEditor/JavascriptEditorViewport.cpp]: This function seems to handle draw behavior for specific use case. I think there should be some functions which exposes FCanvas*Item accessible to BP.

Comment (arthursw) [reply Source/JavascriptEditor/JavascriptEditorViewport.cpp]: Yes, I fully agree, I updated my pull request.

I exposed the `FCanvas` as a `UCanvas` like in `JavascriptEdMode`.

For this purpose, I moved `GetCanvasByName()` from `JavascriptEdMode.cpp` to `JavascriptEditorLibrary.h` to be able to access it from `JavascriptEdMode.cpp` and `JavascriptEditorViewport.cpp`.

Tell me if there is a better solution.

Comment (arthursw) [reply Source/JavascriptEditor/JavascriptEditorViewport.cpp]: I forgot to make some changes... Working on it...

**Comment by nakosung** [comment]
It is recommended to hold UObject references by FGCObject. Could you transform your code not to call AddToRoot?

**Comment by arthursw** [comment]
Ok, I did something, I'm not sure that it is a good solution though.

I added 

```
UProperty()
UCanvas *Canvas;
```

to `UJavascriptEditorViewport` (I moved `GetCanvasByName()` back to `JavascriptEdMode.cpp`).

Instead, I could override `FEditorViewportClient::AddReferencedObjects()` for the `FJavascriptEditorViewportClient` (in `JavascriptEditorViewport.cpp`), maybe it is a better solution?

In `JavascriptEditorViewport.cpp`:

```
    virtual void AddReferencedObjects(FReferenceCollector& Collector) override
    {
        FEditorViewportClient::AddReferencedObjects(Collector);
        Collector.AddReferencedObject(CanvasObject);     // Add a reference to the UCanvas object
    }
```

(In the same way, I could also add `Collector.AddReferencedObject(CanvasObject);` in `FJavascriptEdModeInstance::AddReferencedObjects()` in `JavascriptEdMode.cpp`, and remove `CanvasObject->AddToRoot()` in `GetCanvasByName()`)

I also fixed some bugs in `UJavascriptEditorViewport::DeprojectScreenToWorld()`and `UJavascriptEditorViewport::ProjectWorldToScreen()`, and removed `FIntRect ViewRect;`and `FViewMatrices ViewMatrices;` from `UJavascriptEditorViewport`.

**Comment by nakosung** [comment]
You may create a private object which inherits from FGCObject and override AddReferencedObjects. In this way, GC ref management can be done in right way. :)

I think uproperty only for gc isn't a good way.

**Comment by arthursw** [comment]
Ok, but why it is better to create a private object which inherits from FGCObject than to override `FEditorViewportClient::AddReferencedObjects()`? (FEditorViewportClient is a FGCObject)

**Comment by arthursw** [comment]
Is it better now?

**Comment by nakosung** [comment]
Thanks, it is looking good to me!

**Pull request merged**